sdk/src/main/java/com/google/cloud/dataflow/sdk/util/common/Counter.java

```diff
@@ -418,17 +464,58 @@ public boolean isCompatibleWith(Counter<?> that) {
 
   //////////////////////////////////////////////////////////////////////////////
 
-  /** The name of this counter. */
-  protected final String name;
-
   /** The kind of aggregation function to apply to this counter. */
   protected final AggregationKind kind;
 
   /** The commit state of this counter. **/
   protected final AtomicReference<CommitState> commitState;
 
+  /**
+   * @{link Name} is the full unique name of the counter. Counters must have a
+   * at least a name, and optionally may be associated with a systemName, an originalName,
+   * or both. All of these names together define a counter uniquely.
+   */
+  @AutoValue
+  public abstract static class Name {
+    /** Create a {@link Name} with the originalName on a {@link NameContext}. */
+    public static Name withOriginalName(String name, NameContext context) {
+      return new AutoValue_Counter_Name(name, null,
+          checkNotNull(context.originalName(), "Expected original name in context %s", context));
+    }
+
+    /** Create a {@link Name} with the systemName on a {@link NameContext}. */
+    public static Name withSystemName(String name, NameContext context) {
+      return new AutoValue_Counter_Name(name,
+          checkNotNull(context.systemName(), "Expected system name in context %s", context),
+          null);
+    }
+
+    /** Create a {@link Name} with the originalName on a {@link NameContext}. */
+    public static Name withoutStructure(String name) {
+      return new AutoValue_Counter_Name(name, null, null);
+    }
+
+    /** Returns the non-unique counter name. */
```

Comment:
maybe `counterName()`?

And "Returns the counter name without any associated context."